### PR TITLE
feat(app): App RobotSettings Calibration Tab Pipette Offset Calibrations and Tip Length Calibrations

### LINKED ISSUE #8758: Calibration - Pipette Offset Calibrations
# Overview
The third section on the robot settings page's calibration tab is used to explain pipette offset calibration to the user.
Any pipettes that have been attached and had their offsets calibrated should appear in this section along with the pipette model, serial #, whether the pipette is attached, left or right mount, the tip rack used, and the last calibration date.
Each row in the table also has a 3-dot menu with a set of options outlined in the implementation details section below.

**Scope Update:** Let's defer adding the link to "see all calibration history". We can add them post-5.1

# Design
This artboard shows Pipette offset calibrations complete, No calibrations have not been performed yet, Only deck calibration has been completed, and Calibration health check returned problems: https://www.figma.com/file/bVtCogdwiXj3ndpk4AWXCx/%5Bworking%5D-Milestone-1%3A-Read-and-Execute?node-id=1297%3A90420

# Implementation Details 
Expected 3-dot action menu options for each pipette/row
- All calibrations are complete + pipette is attached: Recalibrate Pipette Offset, Download calibration data, Delete calibration data
- All calibrations are complete + pipette is not attached: Recalibrate Pipette Offset (inactive), Download calibration data, Delete calibration data
- No calibrations are complete: No 3-dot action menu icons should appear. (Deck calibration needs to be run first.)
- Only deck calibration has been completed: Calibrate Tip Length and Pipette Offset

# Acceptance Criteria
- [ ] Currently attached pipettes should be identified with an `*` next to the pipette model and generation 
- [ ] The correct action menu options appear as outlined above in the implementation details
- [ ] The "Recalibrate pipette offset" menu option should kick off the pipette offset calibration workflow
- [ ] The "Delete calibration data" menu option should delete that pipette's saved calibration data

Most recent pipette offset calibration:

- [ ] If pipette offset calibration has been performed, the date and time of the most recent pipette offset calibration should appear correctly
- [ ] If pipette offset calibration has not been performed yet, "Not calibrated yet" should be shown where the date and time normally appear

Errors and warnings:

- [ ] If deck calibration hasn't been performed yet, a red error banner should display "Calibration needed, calibrate deck first" 

- [ ] If deck calibration has been performed yet but pipette offset calibration hasn't, a red error banner should display "Calibration needed" 

- [ ] If calibration health check returns any problems, a yellow warning banner should display "Recalibration recommended"

### PR DESCRIPTION
**Pull request by koji**
# Overview
This will fix #8758 #8760 #8797  #8798

# Changelog
- Added Pipette Offset Calibrations section
- Added Tip Length Calibrations section
- Modify labware function

# Review requests

# Risk assessment
medium because this PR modifies `app/src/assets/labware/findLabware.ts` since jest shows T`ypeError: Cannot read property 'length' of undefined` and this function is used by 5.0.2.

### CODE REVIEW

**Comment by koji** [comment]
- [ ] fix check-js errors
- [x] fix Tip Length issue no data -> not calibrated yet

File: app/src/assets/labware/findLabware.ts [hunk 1]
```diff
@@ -6,22 +6,27 @@ import { getLatestLabwareDef } from './getLabware'
...
+// and handle nullish cases
 function filterLabwareDefinitions(
   namespace: string | null,
   loadName: string | null,
   version: string | null,
   customLabware: LabwareDefinition2[]
 ): LabwareDefinition2 | null {
-  return (
-    head(
+  if (customLabware != null) {
```
Comment (b-cooper) [inline app/src/assets/labware/findLabware.ts]: this doesn't make much sense to me.  The type of this function parameter doesn't include `null | undefined` it's just `LabwareDefinition2[]` so typescript should already be yelling anywhere where `customLabware` is not defined right?

Comment (koji) [reply app/src/assets/labware/findLabware.ts]: you are right. the issue was my test code. I will fix the issue.

**Comment by koji** [comment]
- [x] fix test issue

File: app/src/organisms/Devices/RobotSettings/CalibrationDetails/OverflowMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            {calType === 'pipetteOffset'
+              ? checkPipetteCalibrations != null
+                ? t('overflow_menu_recalibrate_pipette')
+                : t('overflow_menu_calibrate_pipette')
+              : t('overflow_menu_recalibrate_tip_and_pipette')}
+          </MenuItem>
+          <MenuItem onClick={e => handleDownload(calType, e)}>
+            {t('overflow_menu_download_calibration_data')}
+          </MenuItem>
+          {/* TODO 5/6/2021 kj: This is scoped out from 5.1 */}
```
Comment (b-cooper) [inline app/src/organisms/Devices/RobotSettings/CalibrationDetails/OverflowMenu.tsx]:
```suggestion
          {/* TODO 5/6/2021 kj: This is scoped out from 6.0 */}
```

File: app/src/organisms/Devices/RobotSettings/CalibrationDetails/OverflowMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    }
+    setShowOverflowMenu(!showOverflowMenu)
+  }
+
+  const handleOverflowClick: React.MouseEventHandler<HTMLButtonElement> = e => {
+    e.preventDefault()
+    setShowOverflowMenu(!showOverflowMenu)
+  }
+
+  // TODO 5/6/2021 kj: This is scoped out from 5.1
```
Comment (b-cooper) [inline app/src/organisms/Devices/RobotSettings/CalibrationDetails/OverflowMenu.tsx]:
```suggestion
  // TODO 5/6/2021 kj: This is scoped out from 6.0
```

File: app/src/organisms/Devices/RobotSettings/CalibrationDetails/PipetteOffsetCalibrationItems.tsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  formattedPipetteOffsetCalibrations,
+}: PipetteOffsetCalibrationItemsProps): JSX.Element {
+  const { t } = useTranslation('device_settings')
+
+  const customLabwareDefs = useSelector((state: State) => {
+    return getCustomLabwareDefinitions(state)
+  })
+
+  return (
+    <>
```
Comment (b-cooper) [inline app/src/organisms/Devices/RobotSettings/CalibrationDetails/PipetteOffsetCalibrationItems.tsx]: this fragment is unnecessary

File: app/src/organisms/Devices/RobotSettings/CalibrationDetails/TipLengthCalibrationItems.tsx [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+  const checkMountWithAttachedPipettes = (serialNumber: string): Mount => {
+    return ['left', 'right'].find((mount, index) => {
+      const pipette = Object.values(attachedPipettes)[index]
+      return pipette?.id === serialNumber
+    }) as Mount
+  }
+
+  return (
+    <>
```
Comment (b-cooper) [inline app/src/organisms/Devices/RobotSettings/CalibrationDetails/TipLengthCalibrationItems.tsx]: unnecessary fragment

File: app/src/organisms/Devices/RobotSettings/CalibrationDetails/__test__/OverflowMenu.test.tsx [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+import { renderWithProviders, Mount } from '@opentrons/components'
+
+import { i18n } from '../../../../../i18n'
+import { useCalibratePipetteOffset } from '../../../../CalibratePipetteOffset/useCalibratePipetteOffset'
+
+import { OverflowMenu } from '../OverflowMenu'
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+const render = (props: React.ComponentProps<typeof OverflowMenu>) => {
```
Comment (b-cooper) [inline app/src/organisms/Devices/RobotSettings/CalibrationDetails/__test__/OverflowMenu.test.tsx]: can you make the return type `ReturnType<typeof renderWithProviders>` and remove this lint ignore?

File: app/src/organisms/Devices/RobotSettings/CalibrationDetails/__test__/PipetteOffsetCalibrationItems.test.tsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
+import * as React from 'react'
+
+import { renderWithProviders, Mount } from '@opentrons/components'
+
+import { i18n } from '../../../../../i18n'
+import { PipetteOffsetCalibrationItems } from '../PipetteOffsetCalibrationItems'
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```
Comment (b-cooper) [inline app/src/organisms/Devices/RobotSettings/CalibrationDetails/__test__/PipetteOffsetCalibrationItems.test.tsx]: same `ReturnType<typeof renderWithProviders>` and remove this lint ignore?

File: app/src/organisms/Devices/RobotSettings/CalibrationDetails/__test__/TipLengthCalibrationItems.test.tsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  },
+  {
+    tiprack: 'opentrons/opentrons_96_tiprack_1000ul/1',
+    pipette: 'Mock-P2KSV222021011802',
+    lastCalibrated: '2022-11-10T18:15:02',
+    markedBad: false,
+  },
+]
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```
Comment (b-cooper) [inline app/src/organisms/Devices/RobotSettings/CalibrationDetails/__test__/TipLengthCalibrationItems.test.tsx:48]: `ReturnType<typeof renderWithProviders>` ?

File: app/src/organisms/Devices/RobotSettings/__tests__/RobotSettingsCalibration.test.tsx [hunk 9]
```diff
@@ -131,4 +152,53 @@ describe('RobotSettingsCalibration', () => {
...
+      mockPipetteOffsetCalibration1,
+      mockPipetteOffsetCalibration2,
+      mockPipetteOffsetCalibration3,
+      mockPipetteOffsetCalibration4,
+    ])
+    const [{ getByText }] = render()
+    getByText('Pipette Offset calibration recommended')
+  })
+
+  // Tip Length Calibrations this comment will be removed when finish all sections
```
Comment (b-cooper) [inline app/src/organisms/Devices/RobotSettings/__tests__/RobotSettingsCalibration.test.tsx]: can this comment be removed now?

File: app/src/pages/Robots/RobotSettings/DeckCalibrationControl.tsx [hunk 10]
```diff
@@ -273,7 +273,7 @@ export function DeckCalibrationControl(props: Props): JSX.Element {
               },
             ]}
           >
-            <Text>{t('deck_calibration_error_occured')}</Text>
+            <Text>{t('deck_calibration_error_occurred')}</Text>
             <Text>
               {
                 // @ts-expect-error TODO use in operator to protect against non existent error
```
Comment (b-cooper) [inline app/src/pages/Robots/RobotSettings/DeckCalibrationControl.tsx]: now would be a nice time to remove the need for these `ts-expect-error`'s.

should just need a:

```
'error' in createRequest && 
createRequest.error != null &&
RobotApi.getErrorResponseMessage(createRequest.error)
```

**Review by b-cooper — APPROVED** [review]
The changes look good 🍦 .  I dropped a few small comments in.

**Comment by koji** [comment]
- [x] fix feedback
- [x] fix conflicts

**Pull request merged**